### ISSUE TITLE
Remove mandatory dependency from compileJava to jOOQ-codegen-gradle tasks

### ISSUE THREAD

**Issue by ThoSap** [jOOQ/jOOQ#15966]
### Your question

Is there an equivalent way to skip the jOOQ codegen from executing every time in the new jOOQ codegen Gradle plugin, as it is possible in Maven?
https://www.jooq.org/doc/latest/manual/code-generation/codegen-maven/#additional-maven-specific-flags
```xml
<!-- A boolean property (or constant) can be specified here to tell the plugin not to do anything -->
<skip>${skip.jooq.generation}</skip>
```

This was available in the unofficial Gradle plugin with the boolean flag `generateSchemaSourceOnCompilation`
https://github.com/etiennestuder/gradle-jooq-plugin/blob/v8.2.1/README.md#generating-the-jooq-sources

I cannot find anything related in the docs
https://www.jooq.org/doc/latest/manual/code-generation/codegen-gradle/

### jOOQ Version

jOOQ Professional Edition 3.19.0

### Database product and version

Trino 435

### Java Version

OpenJDK 64-Bit Server VM GraalVM CE 21.0.1+12.1 (build 21.0.1+12-jvmci-23.1-b19, mixed mode, sharing)

### OS Version

Microsoft Windows [Version 10.0.22631.2861]

### JDBC driver name and version (include name if unofficial driver)

io.trino:trino-jdbc:435

**Comment by lukaseder** [comment]
This isn't possible yet. Why do you need it?

**Comment by lukaseder** [comment]
Note, one way is to exclude the task from being run, e.g.

```
./gradlew test -x jooqCodegen
```

**Comment by strindberg** [comment]
I am not the creator of this JIRA, but in our project, we use the aforementioned flag `generateSchemaSourceOnCompilation = false` in the unofficial gradle plugin. The reason is mainly because it is quite inconvenient when one is in the middle of a coding task and everything is in a flux, to be required to have all flyway/liquibase files in order to even run a unit test. We have tried both and prefer to only generate the Jooq classes explicitly.

**Comment by lukaseder** [comment]
I can see that, thanks for the explanations, @strindberg.

In Maven, this is exactly what profiles are for. It appears that gradle doesn't have the exact equivalent, but you can emulate it like this:

https://stackoverflow.com/q/40659986/521799

While there's always room for escape hatches like `skip` flags, I feel that the skip flag is just a hack, and there must be a *much* better way to enable/disable tasks by default in a build system. I'm just getting started with gradle, so I don't know the best answer yet, but again, surely, there must be something equivalent to Maven's profiles?

**Comment by strindberg** [comment]
Without having looked at the Jooq gradle plugin's source code, I assume that the jOOQ codegen task is registered as a dependency on the build task. This is of course a sensible default. However, in my experience, it's a bit awkward to break this dependency in a build file, once the dependency is declared by a plugin. 

Thus, for us, a flag just as in the unofficial plugin is exactly what we need.

**Comment by lukaseder** [comment]
> However, in my experience, it's a bit awkward to break this dependency in a build file, once the dependency is declared by a plugin.

Can you elaborate on this experience, please?

**Comment by strindberg** [comment]
OK, first, to re-iterate our needs: in our project we want the generate task to be available, but only run when explicitly specified, i.e. we want the default (no generation) to require no extra flag for our developers. In the unofficial plugin, we obtain this by setting the flag mentioned above, and then run `./gradlew generateJooq`when we want to re-generate the jooq sources.

As to your question, I have actually been experimenting with getting the new official plugin to work with our needs, and I can not find a way to "undo" what the plugin does, such that no extra parameters are necessary on the command line: I find no way of disentangle the `build` task from the `generate` task, neither by specifying that generate is NOT a dependency to build, nor by removing the output from the generate task from the input to the build task.

In short, there might be a way, but I haven't found it. A flag would certainly be convenient for us.

**Comment by lukaseder** [comment]
But have you considered an approach similar to Maven's profiles?
https://docs.gradle.org/current/userguide/migrating_from_maven.html#migmvn:profiles_and_properties

It seems so much cleaner than a hacky skip flag!

**Comment by edvinasstaupas** [comment]
In my opinion, jOOQ codegen task should be by default singular and started explicitly without dependencies on other tasks. It isn't related to other tasks, as it only generates part of the code. If a person wants to both generate code and then for example build the project, they can run `gradle jooqCodegen build`.

**Comment by lukaseder** [comment]
Next time, it would be great to provide fundamental feedback *before* the release ;-)

**Comment by lukaseder** [comment]
Anyway, if someone can please point me to an authoritative TL;DR style tutorial on what people expect when configuring things in their `build.gradle` files, that would be great.

I've never used gradle, tried to stay clear of it for all my professional life. I personally find Maven much cleaner (though incremental builds are nice).

I don't understand the purpose of adding a plugin to a build's "top level" (or whatever it's called), and then expect it *not* to be applied. If application is optional, then there are profiles in Maven. None of you have acknowledged my profile suggestions thus far. Maybe, the concept of profiles is completely dumb in the gradle world? I don't know. I find them perfect:

- Plugins are applied by default
- Opt-in plugins can be moved to profiles

@edvinasstaupas You seem to suggest that *every* silly gradle plugin out there needs to make these decisions on behalf of users? And then, users come and worry about how to hack around the default, because it doesn't suit them? Is that how things work in gradle? Clearly, I must be missing the bigger picture here.

**Comment by strindberg** [comment]
I don't know exactly to whom you addressed your comment about feedback before the release. In general I follow jooq's development with great interest, and would have been happy to test these things out earlier if that would have helped. But the gradle plugin was developed rather quickly just before this release, was it not? Apart from the gradle plugin, which is a bit of  a special case, I usually find I need a bit of time digesting the new features presented in each release, and I have rarely been able to get things up and running quickly enough.

Anyway, the short answer to your Maven suggestion is that imitating Maven is the absolutely last thing we would consider doing. I have avoided Maven for more than 10 years now. I find it absolutely revolting and all the thinking surrounding it. I have rarely used a more patronizing tool. I will never forget the first time I tried Maven and it threw a NullPointerException because there were no tests in my example project. Apparently the Maven authors were so convinced of the need of tests that had not even bothered to test for their existence.

I can't say you have really explained what you find hacky about the flag. Our use case is that we want the task to be available, but not run in every build.  It's probably a good idea to have it run by default, but then we need a way to turn it off. 

I interpreted the stated goal of this plugin to eventually replace the unofficial one. I don't think I was alone in expecting this to be a more or less drop in replacement of the previous plugin.

**Comment by strindberg** [comment]
One more thing: the flyway plugin has its own issues, but is perhaps a good example of when a plugin works well by offerings tasks that can be run without making them mandatory: it doesn't really make sense to run all the flyway tasks at every build, but it's still convenient to have them available in your gradle script.

I am not enough of a gradle expert to offer you a plugin style guide, but as a gradle user I would expect different behavior from different types of plugins: I don't want all flyway tasks to run at every build, but on the other hand, once I activate the kotlin plugin I expect it to always compile my code without me specifying it. Depends on how central to my build the tool is, really. In my mind, jooq falls in the middle of these two example.

**Comment by edvinasstaupas** [comment]
I haven't used gradle profiles, so I can't comment on it. Attaching task to the original build task is quite easy:
```
build {
    dependsOn 'generateOpenApiDocs'
    mustRunAfter 'generateOpenApiDocs'
}
```
then running `gradle build` or `gradle assemble` also runs `generateOpenApiDocs` before `build` but after `compile`. 
I am using `id 'org.springdoc.openapi-gradle-plugin' version "1.6.0"` for the example.

**Comment by lukaseder** [comment]
> I don't know exactly to whom you addressed your comment about feedback before the release. In general I follow jooq's development with great interest, and would have been happy to test these things out earlier if that would have helped. But the gradle plugin was developed rather quickly just before this release, was it not?

I'm just venting, because this stuff is so hard to get right. Gradle's documentation is huge, quickly loses itself in irrelevant (to me) details, and much of it is about how superior it is to Maven. Every tutorial or Stack Overflow question hacks around things by programmatically doing arbitrary changes, so it's never really clear what The Right Way&trade; is supposed to be (unlike the patronising Maven, which has only one Right Way).

Much of the plugin was developed in July 2023. The recent improvements were about getting the DSL right (as opposed to letting users call the programmatic API, which is a bit ugly), but I haven't changed anything fundamental anymore since July.

Anyway, it's probably not too late to change it incompatibly again in the 3.19.2 patch release.

> Anyway, the short answer to your Maven suggestion is that imitating Maven is the absolutely last thing we would consider doing. I have avoided Maven for more than 10 years now. I find it absolutely revolting and all the thinking surrounding it. I have rarely used a more patronizing tool. I will never forget the first time I tried Maven and it threw a NullPointerException because there were no tests in my example project. Apparently the Maven authors were so convinced of the need of tests that had not even bothered to test for their existence.

I once thought like you, but gave in to the patronising approach, and found that it is really really good. A NPE is just a bug, so it's not really a strategic argument against it, I'd say? I recently ran into NPEs in javac or even the javadoc tools, yet, I won't throw them out the window? But anyway, it's moot to discuss preferences here. Obviously, most gradle users prefer gradle over maven.

All I want to know is what is The Right Way&trade; for jOOQ-codegen-gradle to behave.

> I can't say you have really explained what you find hacky about the flag. Our use case is that we want the task to be available, but not run in every build. It's probably a good idea to have it run by default, but then we need a way to turn it off.

My point is, perhaps the auto-activation of the task is what's wrong here (which is what @edvinasstaupas suggests). If gradle doesn't have profiles, then I guess my mental model of a plugin auto-applying to a build as soon as it's configured at the top level is wrong. And if it's wrong, then the flag working around it is also wrong (i.e. a hack).

Why do you think it's a good idea to have it run by default?

> I interpreted the stated goal of this plugin to eventually replace the unofficial one. I don't think I was alone in expecting this to be a more or less drop in replacement of the previous plugin.

Yes, but this one doesn't have to do everything exactly like the other. The main reason for this plugin is:

- The other isn't versioned like jOOQ
- The other doesn't support the commercial editions very well

But!

- This one won't focus on the Gradle Enterprise Edition or other highly Gradle specific use-cases

So, no. Eventually, they won't be drop in replacements of one another.

> One more thing: the flyway plugin has its own issues, but is perhaps a good example of when a plugin works well by offerings tasks that can be run without making them mandatory: it doesn't really make sense to run all the flyway tasks at every build, but it's still convenient to have them available in your gradle script.

I'm sorry if I have to invoke my Maven-fu again here, but in Maven, tasks are goals, and a plugin is configured with a specific goal, and the fact whether the plugin is executed is based purely on where the configuration is located:

- Top level = auto execution
- Profile = opt-in execution

Obviously, no one wants to auto-execute all the goals of a plugin in Maven either, but the fact that a plugin is configured at some location strongly hints at the intent.

Now, I don't want to replicate Maven here, but I want to do the right thing. And I'm starting to agree with @edvinasstaupas. Auto-execution maybe isn't the right thing.

> I am not enough of a gradle expert to offer you a plugin style guide, but as a gradle user I would expect different behavior from different types of plugins: I don't want all flyway tasks to run at every build, but on the other hand, once I activate the kotlin plugin I expect it to always compile my code without me specifying it. Depends on how central to my build the tool is, really. In my mind, jooq falls in the middle of these two example.

And that is precisely why I think the Maven approach is so intuitive. Because you'll configure each plugin according your needs, instead of having to work around default behaviour, when it's wrong.

**Comment by strindberg** [comment]
My anecdote about the NullPointerException is relevant in my mind because every other experience I have had with Maven have been exactly the same: some randomly chosen Maven developer's preference is more important than the needs we have in our project. But let's leave that discussion: neither of us will switch preference.

Anyway, I realize my comment about auto-activation was misleading in this context: I just wanted to express the idea that you had not made a crazy choice: there is a certain convenience if things just work.

But I do think we might be agreement here: In gradle, adding a task as a dependency is very easy (as @edvinasstaupas has demonstrated), but it is difficult to break existing dependency chains. So either we would need this flag we are discussing, or the jooq tasks should not automatically be run.

**Comment by edvinasstaupas** [comment]
Everyone can change when and how they want `jooqCodegen` to run if it by default is independent. 
Current behaviour (if I understand correctly) could be implemented like this:
```
compileJava {
    dependsOn 'jooqCodegen'
}
```

**Comment by lukaseder** [comment]
OK, thanks for your feedback, folks. I won't rush any decisions here. I'll have to invest more time looking at how other code generation utilities (e.g. xjc) do things. This won't happen before mid January 2024, though.

**Comment by ThoSap** [comment]
With the current state, it is also not possible to configure the Flyway migration to run before the `jooqCodegen` or `jooqCodegenMain` tasks.

```groovy
// configure jOOQ task such that it only executes when something has changed that potentially affects the generated JOOQ sources
// - the jOOQ configuration has changed (Jdbc, Generator, Strategy, etc.)
// - the classpath used to execute the jOOQ generation tool has changed (jOOQ library, database driver, strategy classes, etc.)
// - the schema files from which the schema is generated and which is used by jOOQ to generate the sources have changed (scripts added, modified, etc.)
tasks.named('jooqCodegen' /* or 'jooqCodegenMain' */).configure {
    // ensure database schema has been prepared by Flyway before generating the jOOQ sources
    dependsOn tasks.named('flywayMigrate')

    // declare Flyway migration scripts as inputs on the jOOQ task
    inputs.files(fileTree('src/main/resources/db'))
        .withPropertyName('migrations')
        .withPathSensitivity(PathSensitivity.RELATIVE)

    // make jOOQ task participate in incremental builds (and build caching)
    allInputsDeclared = true
}
```

With this, I get the error:
```java
Build file 'C:\Users\ThoSap\IdeaProjects\my-project\build.gradle' line: 370

A problem occurred evaluating root project 'my-project'.
> Task with name 'jooqCodegen' /* or 'jooqCodegenMain' */ not found in root project 'my-project'.
```

See the following example with the unofficial plugin:
https://github.com/etiennestuder/gradle-jooq-plugin/blob/v9.0/example/configure_jooq_with_flyway/build.gradle#L54-L69

**Comment by lukaseder** [comment]
@ThoSap I believe that's a different issue: https://github.com/jOOQ/jOOQ/issues/15974

**Comment by lukaseder** [comment]
Thanks for your patience folks. 

I read up a bit more on best practices, plus investigated why I had implemented things the way I did in the first place. I can't really say for sure anymore... most likely reason is that I misunderstood something when the typical error message popped up:

> Reason: Task ':compileJava' uses this output of task ':jooqCodegenMain' without declaring an explicit or implicit dependency.

And due to #15974, I couldn't fix this error from within the `build.gradle.kts` file, so I mistakenly concluded that the fix must go into the plugin. Obviously, that's not correct, nor was it complete, given that only `compileJava` was "fixed", but not e.g. kotlin compilation, etc.

So, I conclude again, that I will remove the mandatory dependency from `compileJava` on the `jOOQ-codegen-gradle` tasks. I think that's a better solution than introducing a flag to skip this dependency.

I will add more tests over the next days, and then probably publish a breaking change in 3.19.2 next week. The breaking part of the change is acceptable, given that the status quo is probably not what most people want.

**Comment by Spikhalskiy** [comment]
I want to provide our feedback in support of the decision made above. We also use the unofficial plugin at the moment, we set the flag to not regenerate the sources automatically and prefer to trigger `./gradlew generateJooq` explicitly on demand. It allows us to play with a DB migration script freely until the desired DB layout is achieved without having to deal with the correspondent interim changes in the generated POJOs.

**Comment by lukaseder** [comment]
Fixed in versions:

- 3.20.0
- 3.19.2 (https://github.com/jOOQ/jOOQ/issues/16058)

**Comment by pareekdevanshu** [comment]
Hi @lukaseder, any plans for fixing it for kotlin compilation as well?
